src/OffloadGPULoops.cpp

```diff
@@ -0,0 +1,345 @@
+
+        // get the actual name of the generated kernel for this loop
+        kernel_name = gpu_codegen->get_current_kernel_name();
+        debug(2) << "Compiled launch to kernel \"" << kernel_name << "\"\n";
+
+        bool runtime_run_takes_types = gpu_codegen->kernel_run_takes_types();
+        Type target_size_t_type = target.bits == 32 ? Int(32) : Int(64);
+
+        vector<Expr> args, arg_types_or_sizes, arg_is_buffer;
+        for (const DeviceArgument &i : closure_args) {
+            Expr val;
+            if (i.is_buffer) {
+                val = Variable::make(Handle(), i.name + ".buffer");
+            } else {
+                val = Variable::make(i.type, i.name);
+                val = Call::make(type_of<void *>(), Call::make_struct, {val}, Call::Intrinsic);
+            }
+            args.push_back(val);
+
+            if (runtime_run_takes_types) {
+                arg_types_or_sizes.push_back(make_type_arg(i.type.with_lanes(1)));
+            } else {
+                arg_types_or_sizes.push_back(cast(target_size_t_type, i.is_buffer ? 8 : i.type.bytes()));
+            }
+
+            arg_is_buffer.push_back(cast<uint8_t>(i.is_buffer));
+        }
+
+        // nullptr-terminate the lists
+        args.push_back(reinterpret(Handle(), cast<uint64_t>(0)));
+        if (runtime_run_takes_types) {
+            internal_assert(halide_type_int == 0);
+            arg_types_or_sizes.push_back(make_type_arg(Type(halide_type_int, 0, 0)));
+        } else {
+            arg_types_or_sizes.push_back(cast(target_size_t_type, 0));
+        }
+        arg_is_buffer.push_back(cast<uint8_t>(0));
+
+        // TODO: only three dimensions can be passed to
+        // cuLaunchKernel. How should we handle blkid[3]?
+        internal_assert(is_const_one(bounds.num_threads[3]) && is_const_one(bounds.num_blocks[3]))
+            << bounds.num_threads[3] << ", " << bounds.num_blocks[3] << "\n";
+        debug(3) << "bounds.num_blocks[0] = " << bounds.num_blocks[0] << "\n";
+        debug(3) << "bounds.num_blocks[1] = " << bounds.num_blocks[1] << "\n";
+        debug(3) << "bounds.num_blocks[2] = " << bounds.num_blocks[2] << "\n";
+        debug(3) << "bounds.num_threads[0] = " << bounds.num_threads[0] << "\n";
+        debug(3) << "bounds.num_threads[1] = " << bounds.num_threads[1] << "\n";
+        debug(3) << "bounds.num_threads[2] = " << bounds.num_threads[2] << "\n";
+
+        string api_unique_name = gpu_codegen->api_unique_name();
+        vector<Expr> run_args = {
+            module_state(api_unique_name),
+            kernel_name,
+            Expr(bounds.num_blocks[0]),
+            Expr(bounds.num_blocks[1]),
+            Expr(bounds.num_blocks[2]),
+            Expr(bounds.num_threads[0]),
+            Expr(bounds.num_threads[1]),
+            Expr(bounds.num_threads[2]),
+            Expr(bounds.shared_mem_size),
+            Call::make(Handle(), Call::make_struct, arg_types_or_sizes, Call::Intrinsic),
```

Comment:
I'm confused about this. It looks like it's a pointer to an array of pointers (because make_type_arg returns a pointer), but the runtime expects a pointer to an array of types

Reply:
I changed this, but I don't have any working OpenGL API on my machine that takes types, so I can't test it... we'll have to see what the buildbots say.